### PR TITLE
feat: refactored arrange_spans to avoid integer overflows for very large huxtables

### PR DESCRIPTION
**Pull request by kpagacz**
I also run `styler` on this file. Let me know if you want me to revert the styler changes.

For details see https://github.com/hughjonesd/huxtable/issues/256

### CODE REVIEW

**Review by hughjonesd — COMMENTED** [review]
OK, so questions and comments:
1. The PR fixes things by just avoiding large numbers in `merge_sets`, right? Couldn't it also just work e.g. by using something like `paste0(row, "/", col)` to create the unique cells? I.e. are the numbers used for anything but differentiating different cellspans?
2. This is already very hairy code and rarely touched. It needs to be as easy to understand as possible. 
*  Do we need the 3D array structure and all the applies? This strikes me as making things too hard to read.
* Comments need to be synced. 
Basically, I can't only merge this if I can read it and understand it; otherwise it will set me up for future pain.

**Comment by hughjonesd** [comment]
Also, yes, please revert styler. I agree the codebase needs styling... but doing just one file is a distraction here.

File: R/subset-extract-internal.R [hunk 1]
```diff
@@ -223,31 +224,39 @@ arrange_spans <- function (
...
+      # go down from i/j until you find a different merge set,
+      # or you have seen ALL the row_numbers/col_numbers
+      end_row <- min(which(
+        row_seq >= i &
+          apply(merge_sets[, j, ], MARGIN = 1, FUN = \(coords) any(coords != ms))
+      )) - 1
+      end_col <- min(which(
+        col_seq >= j &
+          apply(merge_sets[i, , ], MARGIN = 1, FUN = \(coords) any(coords != ms))
+      )) - 1
```
Comment (hughjonesd) [inline R/subset-extract-internal.R:240]: This is very difficult... though the existing code is hardly easy....

File: R/subset-extract-internal.R [hunk 2]
```diff
@@ -200,16 +200,17 @@ arrange_spans <- function (
...
   row_number <- row_number[rows, cols, drop = FALSE]
   col_number <- col_number[rows, cols, drop = FALSE]
 
   # == add 0s to ensure end_row/col is ok when a span goes to the end of the row
-  merge_sets <- rbind(merge_sets, 0)
-  merge_sets <- cbind(merge_sets, 0)
+  new_merge_sets_dims <- c(dim(merge_sets)[1:2] + 1, dim(merge_sets)[3])
+  new_merge_sets <- array(0, dim = new_merge_sets_dims)
+  new_merge_sets[-new_merge_sets_dims[1], -new_merge_sets_dims[2], ] <- merge_sets
+  merge_sets <- new_merge_sets
```
Comment (hughjonesd) [inline R/subset-extract-internal.R:213]: The array structure definitely makes this harder to understand

**Comment by kpagacz** [comment]
> Also, yes, please revert styler. I agree the codebase needs styling... but doing just one file is a distraction here.

Done, reverted.

Regarding some other solutions - I quite liked the `as.double` if it actually works. My worry there was that double precision is bad for large numbers and the equality check could potentially yield false positives, but I checked and the first pair of integers that evaluates as the same starts near 10^50, so I think we are safe here.

Otherwise, I can add some comments or whatever you want me to add to make it more digestible. When developing I thought about it as instead of having one int32 to identify a cell, we now have a pair of two int32 that live inside `merge_sets[i,j]` instead of one int32.

**Comment by kpagacz** [comment]
I am closing this since you seemingly solved it with `as.double`...

**Comment by hughjonesd** [comment]
You obviously delved deeply into huxtable internals here! Thank you for the accurate bug report which made the fix (hopefully) so easy.
